### PR TITLE
Do not load whole jar file when --function is given

### PR DESCRIPTION
**Pull request by antlechner**
Only start the analysis of loading classes at the main class, i.e. the class that contains the entry method.

We currently load a lot of classes that we don't need when jbmc is given a jar file as an argument. This change prevents that, and running on a .jar should get performance close to running on a .class.

I don't think we have any tests for large jar files. Locally I got very promising results with these changes, and I will try running them on some company benchmarks. If the results from those runs look good, I will remove the "do not merge" label and suggest we merge it, and worry later about how to write good tests for large jar files (but we should definitely do that too).

### CODE REVIEW

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 1]
```diff
@@ -213,6 +213,13 @@ bool java_bytecode_languaget::parse(
       java_cp_include_files);
     if(config.java.main_class.empty())
     {
+      const std::string &entry_method = config.main;
+      // If we have an entry method, we can derive a main class.
+      if(entry_method != "")
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: use `!entry_method.empty()` here

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 2]
```diff
@@ -213,6 +213,13 @@ bool java_bytecode_languaget::parse(
...
+      {
+        const auto last_dot_position = entry_method.find_last_of('.');
+        main_class = entry_method.substr(0, last_dot_position);
+      }
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: maybe reading the manifest should be in an `else` branch. As it is currently, a main entry in the manifest would override the setting from `--function`

**Review by mgudemann — CHANGES_REQUESTED** [review]

File: jbmc/src/java_bytecode/java_bytecode_language.cpp (hunk 1 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/java_bytecode_language.cpp]: Done

File: jbmc/src/java_bytecode/java_bytecode_language.cpp (hunk 2 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/java_bytecode_language.cpp]: Done

**Review by smowton — COMMENTED** [review]
Add a regression test, using `--verbosity 10` so you can observe the "loading class from JAR" lines -- you can for example check that a class present in a jar but not referenced from the entry point doesn't get loaded.

**Review by mgudemann — APPROVED** [review]
LGTM

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: f4f9ca0b4487e404bcbe3b73db94b7f65c02be4a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84990112

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 06481d7814c4aefb23b05a37c6ed4ebb47b92b52).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85051606

**Review by allredj — COMMENTED** [review]
Would it be useful to document that behaviour somewhere in the CProver manual?

**Comment by antlechner** [comment]
@smowton I added a test as you suggested.
@allredj I added some doxygen above `java_bytecode_languaget::parse` and also updated the output of `jbmc --help` to explain that jbmc can run on .class and .jar files. @peterschrammel said that the output `--janalyzer --help` should be updated accordingly, so I made sure the two are the same.
Running this on benchmarks looks promising (not a huge improvement but it definitely hasn't gotten worse), so this is ready for re-review now.

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 3]
```diff
@@ -207,12 +207,22 @@ bool java_bytecode_languaget::parse(
...
+      else
+      {
+        auto manifest = java_class_loader.jar_pool(path).get_manifest();
+        std::string manifest_main_class = manifest["Main-Class"];
 
-      // if the manifest declares a Main-Class line, we got a main class
-      if(manifest_main_class!="")
-        main_class=manifest_main_class;
+        // if the manifest declares a Main-Class line, we got a main class
+        if(manifest_main_class != "")
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: ! .empty()

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 4]
```diff
@@ -733,7 +733,14 @@ void janalyzer_parse_optionst::help()
...
     " janalyzer [-?] [-h] [--help] show help\n"
-    " janalyzer class              name of class to be checked\n"
+    " janalyzer class              name of class or JAR file to be checked\n"
+    "                              In the case of a JAR file, if a main class"
+    " can be inferred\n"
+    "                              from --main-class, --function or the JAR"
+    " manifest (checked in\n"
+    "                              this order), the behavior is the same as"
+    " running janalyzer\n"
+    "                              on the corresponding class file."
```
Comment (allredj) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: :pick: line wrapping is weird

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 5]
```diff
@@ -1022,7 +1022,14 @@ void jbmc_parse_optionst::help()
...
     "\n"
     " jbmc [-?] [-h] [--help]      show help\n"
-    " jbmc class                   name of class to be checked\n"
+    " jbmc class                   name of class or JAR file to be checked\n"
+    "                              In the case of a JAR file, if a main class"
+    " can be inferred\n"
+    "                              from --main-class, --function or the JAR"
+    " manifest (checked in\n"
+    "                              this order), the behavior is the same as"
+    " running jbmc on the\n"
```
Comment (allredj) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: :pick: line wrapping is weird

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 6]
```diff
@@ -733,7 +733,14 @@ void janalyzer_parse_optionst::help()
     "Usage:                       Purpose:\n"
     "\n"
     " janalyzer [-?] [-h] [--help] show help\n"
-    " janalyzer class              name of class to be checked\n"
+    " janalyzer class              name of class or JAR file to be checked\n"
+    "                              In the case of a JAR file, if a main class"
+    " can be inferred\n"
```
Comment (peterschrammel) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: The help text should be broken after 80 chars and properly aligned.

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 4 above)
Comment (allredj) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: Does clang let you indent the (for instance) `" running janalyzer` line?

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 7]
```diff
@@ -1022,7 +1022,14 @@ void jbmc_parse_optionst::help()
     "Usage:                       Purpose:\n"
     "\n"
     " jbmc [-?] [-h] [--help]      show help\n"
-    " jbmc class                   name of class to be checked\n"
+    " jbmc class                   name of class or JAR file to be checked\n"
+    "                              In the case of a JAR file, if a main class"
+    " can be inferred\n"
+    "                              from --main-class, --function or the JAR"
```
Comment (allredj) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: :pick: Oxford comma after `--function` helps understanding I think.

**Review by allredj — COMMENTED** [review]
Do you have a test-gen update?

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 4 above)
Comment (antlechner) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: clang isn't a problem, CLion is doing the weird line wrapping thing. What was already in the file doesn't look much nicer but I can try to format the new lines in the same way using vim.

**Comment by antlechner** [comment]
@allredj Yes, it's very old and you approved it a long time ago already. :)
https://github.com/diffblue/test-gen/pull/2181

File: jbmc/src/java_bytecode/java_bytecode_language.cpp (hunk 3 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/java_bytecode_language.cpp]: Done

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 6 above)
Comment (antlechner) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: I think the formatting is correct now (as it was done below in the same file).

File: jbmc/src/jbmc/jbmc_parse_options.cpp (hunk 5 above)
Comment (antlechner) [reply jbmc/src/jbmc/jbmc_parse_options.cpp]: Fixed (I think - it's similar to what was done below in the same file).

File: jbmc/src/jbmc/jbmc_parse_options.cpp (hunk 7 above)
Comment (antlechner) [reply jbmc/src/jbmc/jbmc_parse_options.cpp]: Done

**Review by allredj — APPROVED** [review]
Looks good!

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 3c9b025c265b90bae5cb9e00e4f69bde50bbeb61).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85303843
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 117457d4de8ec2972c8269f9636176cff8fb31a8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85320651

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 7d7424f081f65d99249a4989b1b781d75c21acec).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85559729

**Pull request merged**